File: SortVision/public/code/heap/julia/heapSort.jl
# Heap Sort (in-place, max-heap)
# Time: Best/Avg/Worst O(n log n), Space: O(1)
function heap_sort!(arr::Vector{Int})
    n = length(arr)

    for i in div(n, 2):-1:1
        heapify!(arr, n, i)
    end

    for i in n:-1:2
        arr[1], arr[i] = arr[i], arr[1]
        heapify!(arr, i - 1, 1)
    end
    return arr
end

function heapify!(arr::Vector{Int}, n::Int, i::Int)
    largest = i
    left = 2 * i
    right = 2 * i + 1

    if left <= n && arr[left] > arr[largest]
        largest = left
    end
    if right <= n && arr[right] > arr[largest]
        largest = right
    end

    if largest != i
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify!(arr, n, largest)
    end
end

arr = [64, 25, 12, 22, 11]
println(heap_sort!(arr))

File: SortVision/public/code/heap/dart/heapSort.dart
class HeapSort {
  // Time: Best/Avg/Worst O(n log n), Space: O(1)
  static void sort(List<int> arr) {
    final n = arr.length;
    for (int i = n ~/ 2 - 1; i >= 0; i--) {
      _heapify(arr, n, i);
    }
    for (int i = n - 1; i > 0; i--) {
      final t = arr[0];
      arr[0] = arr[i];
      arr[i] = t;
      _heapify(arr, i, 0);
    }
  }

  static void _heapify(List<int> arr, int n, int i) {
    int largest = i;
    final left = 2 * i + 1;
    final right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest]) largest = left;
    if (right < n && arr[right] > arr[largest]) largest = right;

    if (largest != i) {
      final t = arr[i];
      arr[i] = arr[largest];
      arr[largest] = t;
      _heapify(arr, n, largest);
    }
  }
}

void main() {
  final arr = [64, 25, 12, 22, 11];
  HeapSort.sort(arr);
  print(arr);
}
